File: legacy_web/assets/js/services/authService.js

/* 
   authService.js - Authentication service
   Handles login, registration, and user profile operations with Supabase
   Supports two user roles: 'user' (deportista) and 'admin_cancha' (administrador)
    */

/* 
   LOGIN - Authenticate user with email/password
 
   @param {string} email    - User's email address
   @param {string} password - User's password
   @returns {object}        - User data (nombre, apellido, rol)
   @throws {Error}          - If authentication fails
   */
async function loginUser(email, password) {
  // Sign in with Supabase using email and password
  const { data: sessionData, error } =
    await supabaseClient.auth.signInWithPassword({
      email,
      password
    });

  // If there's an error, throw it to be handled by the caller
  if (error) throw error;

  // Fetch additional user data from the 'usuarios' table — include rol
  const { data: usuario, error: dbError } =
    await supabaseClient
      .from('usuarios')
      .select('nombre, apellido, rol')  // Include rol for redirect logic
      .eq('id', sessionData.user.id)
      .single();

  // If database error, throw custom error message
  if (dbError) throw new Error('No se encontraron datos del usuario.');

  // Return user profile data (includes rol)
  return usuario;
}

/* 
   REGISTER - Create new user account
   @param {string} email    - User's email address
   @param {string} password - User's chosen password
   @param {string} nombre   - User's first name
   @param {string} apellido - User's last name
   @returns {object}        - Created user object
   @throws {Error}          - If registration fails
   */
async function registerUserAuth(email, password, nombre, apellido) {
  // Sign up new user with Supabase Auth
  const { data, error } = await supabaseClient.auth.signUp({
    email,
    password,
    options: {
      // Store additional user metadata in Auth
      data: { nombre, apellido }
    }
  });

  // If there's an error, throw it
  if (error) throw error;

  // Return the created user object
  return data.user;
}

/* 
   INSERT PROFILE - Create user profile in database
   
   @param {string} userId - The authenticated user's ID
   @param {object} data   - Profile data { name, lastname, phone, email }
   @returns {void}
   @throws {Error}        - If insert fails
   */
async function insertUserProfile(userId, data) {
  // Insert new row into usuarios table
  const { error } = await supabaseClient
    .from('usuarios')
    .insert([{
      id: userId,                         // Link to auth user ID
      nombre: data.name,                  // First name
      apellido: data.lastname,            // Last name
      telefono: data.phone || null,       // Phone (optional)
      correo_electronico: data.email,     // Email
      rol: data.rol || 'user',            // Role: 'user' | 'admin_cancha'
    }]);

  // If there's an error, throw it
  if (error) throw error;
}

/* 
   GET USER ROLE - Fetch the current authenticated user's role

   @returns {string} - 'user' | 'admin_cancha' | null
   @description - Queries usuarios table for the role of the active session user
   */
async function getUserRole() {
  // Get current session
  const { data } = await supabaseClient.auth.getSession();
  const userId = data?.session?.user?.id;
  if (!userId) return null;

  // Query rol from usuarios table
  const { data: row, error } = await supabaseClient
    .from('usuarios')
    .select('rol')
    .eq('id', userId)
    .single();

  if (error) return null;
  return row?.rol ?? 'user';
}

// Expose functions globally for use in other scripts
window.loginUser         = loginUser;
window.registerUserAuth  = registerUserAuth;
window.insertUserProfile = insertUserProfile;
window.getUserRole       = getUserRole;

File: lib/services/auth_service.dart
import 'package:supabase_flutter/supabase_flutter.dart';
import '../models/usuario.dart';

class AuthService {
  final SupabaseClient _client = Supabase.instance.client;

  // Patrón Singleton
  AuthService._internal();
  static final AuthService instance = AuthService._internal();

  User? get currentUser => _client.auth.currentUser;
  Session? get currentSession => _client.auth.currentSession;

  /// Inicia sesión con email y contraseña, luego recupera el perfil del usuario
  Future<Usuario> loginUser(String email, String password) async {
    final response = await _client.auth.signInWithPassword(
      email: email,
      password: password,
    );

    if (response.user == null) {
      throw Exception('Autenticación fallida: el servidor no retornó un usuario.');
    }

    return await getUserProfile(response.user!.id);
  }

  /// Registra un nuevo usuario en Supabase Auth
  Future<User?> registerUserAuth({
    required String email,
    required String password,
    required String nombre,
    required String apellido,
  }) async {
    final response = await _client.auth.signUp(
      email: email,
      password: password,
      data: {
        'nombre': nombre,
        'apellido': apellido,
      },
    );
    return response.user;
  }

  /// Inserta el perfil del usuario en la tabla 'usuarios' de la base de datos
  Future<void> insertUserProfile({
    required String userId,
    required String email,
    required String nombre,
    required String apellido,
    String? telefono,
    required String rol,
  }) async {
    await _client.from('usuarios').insert({
      'id': userId,
      'nombre': nombre,
      'apellido': apellido,
      'telefono': telefono,
      'correo_electronico': email,
      'rol': rol,
    });
  }

  /// Obtiene los detalles de perfil de un usuario por su ID
  Future<Usuario> getUserProfile(String userId) async {
    final data = await _client
        .from('usuarios')
        .select('id, nombre, apellido, telefono, correo_electronico, rol')
        .eq('id', userId)
        .single();
    
    return Usuario.fromJson(data);
  }

  /// Obtiene el rol del usuario actualmente autenticado
  Future<String?> getUserRole() async {
    final userId = _client.auth.currentUser?.id;
    if (userId == null) return null;

    try {
      final data = await _client
          .from('usuarios')
          .select('rol')
          .eq('id', userId)
          .single();
      return data['rol'] as String?;
    } catch (_) {
      return 'user'; // Rol por defecto si no se encuentra
    }
  }

  /// Cierra la sesión
  Future<void> signOut() async {
    await _client.auth.signOut();
  }
}
